src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs

```diff
@@ -0,0 +1,177 @@
+                        throw new InvalidOperationException(Resources.FormatApiController_AttributeRouteRequired(nameof(ApiControllerAttribute)));
+                    }
+
+                    if (_apiBehaviorOptions.EnableModelStateInvalidFilter)
+                    {
+                        Debug.Assert(_apiBehaviorOptions.InvalidModelStateResponseFactory != null);
+                        actionModel.Filters.Add(_modelStateInvalidFilter);
+                    }
+
+                    if (_modelMetadataProvider != null && _apiBehaviorOptions.InferBindingSourcesForParameters)
+                    {
+                        foreach (var parameter in actionModel.Parameters)
+                        {
+                            var bindingSource = InferBindingSourceForParameter(parameter);
+                            if (bindingSource != null)
+                            {
+                                _logger.InferringParameterBindingSource(parameter, bindingSource);
+                                parameter.BindingInfo = new BindingInfo
+                                {
+                                    BindingSource = bindingSource,
+                                };
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Internal for unit testing.
+        internal BindingSource InferBindingSourceForParameter(ParameterModel parameter)
+        {
+            if (parameter.BindingInfo != null)
+            {
+                return null;
+            }
+
+            if (ParameterExistsInAllRoutes(parameter.Action, parameter.ParameterName))
+            {
+                return BindingSource.Path;
+            }
+            else
+            {
+                var parameterMetadata = _modelMetadataProvider.GetMetadataForParameter(parameter.ParameterInfo);
+                if (parameterMetadata != null)
+                {
+                    var bindingSource = parameterMetadata.IsComplexType ?
+                        BindingSource.Body :
+                        BindingSource.Query;
+
+                    return bindingSource;
+                }
+            }
+
+            return null;
+        }
+
+        private bool ParameterExistsInAllRoutes(ActionModel actionModel, string parameterName)
+        {
+            var parameterExistsInSomeRoute = false;
+            foreach (var routeTemplate in GetAllRouteTemplates())
+            {
+                var parsedTemplate = TemplateParser.Parse(routeTemplate);
+                if (parsedTemplate.GetParameter(parameterName) == null)
+                {
+                    return false;
+                }
+
+                // Ensure at least one route exists.
+                parameterExistsInSomeRoute = true;
+            }
+
+            return parameterExistsInSomeRoute;
+
+            IEnumerable<string> GetAllRouteTemplates()
```

Comment:
Is this code a dupe of something else? Should we consolidate this or provide extension methods?

Reply:
It's a clone of https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs#L147. I'll see if I can refactor it to be re-used.